### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by wixoaGit**
Ended up changing a lot more than I originally intended, but I got error handling working in a way I'm reasonably happy with.

I changed how parsing errors are caught. `Error()` now throws a `CompileErrorException` that is caught by `BlockInner()`, `ProcBlockInner()`, and `File()`. When BlockInner and ProcBlockInner catch this exception they skip to the next statement using `LocateNextStatement()`.

There are some rare cases where the parser fails to recover from an error correctly. When this happens `File()` will catch any CompileErrorException, realize it's not at an EOF token, and attempt to skip to the next top-level statement using `LocateNextTopLevel()` while emitting a warning. Cases where this happens should be avoided and fixed in the future.

For now I've removed any printing related to error context. The implementation for this was very parser-specific, while I'd like a future implementation to be more general. This would allow errors emitted by a lexer or visitor to also have context information.
I'd prefer to have this done outside of the error handling pr.

### CODE REVIEW

File: OpenDreamShared/Compiler/DM/DMParser.cs [hunk 1]
```diff
@@ -12,197 +12,156 @@ public partial class DMParser : Parser<Token> {
...
+                    if (!PeekDelimiter() && Current().Type != TokenType.DM_Dedent) {
+                        Error("Expected end of statement");
+                    }
+
+                    _currentPath = oldPath;
+                    return statement;
+                } catch (CompileErrorException) {
+                    _currentPath = oldPath;
+                    throw; //Pass onto BlockInner()
+                }
```
Comment (mumencoder) [inline OpenDreamShared/Compiler/DM/DMParser.cs:163]:
```suggestion
                }
                finally { _currentPath = oldPath; }
```

File: OpenDreamShared/Compiler/DM/DMParser.cs [hunk 2]
```diff
@@ -12,197 +12,156 @@ public partial class DMParser : Parser<Token> {
...
+                    //Empty object definition
+                    if (statement == null) {
+                        statement = new DMASTObjectDefinition(_currentPath, null);
+                    }
+
+                    if (!PeekDelimiter() && Current().Type != TokenType.DM_Dedent) {
+                        Error("Expected end of statement");
+                    }
+
+                    _currentPath = oldPath;
```
Comment (mumencoder) [inline OpenDreamShared/Compiler/DM/DMParser.cs]:
```suggestion

```

File: OpenDreamShared/Compiler/DM/DMParser.cs [hunk 3]
```diff
@@ -12,197 +12,156 @@ public partial class DMParser : Parser<Token> {
...
+                    return statement;
+                } catch (CompileErrorException) {
+                    _currentPath = oldPath;
```
Comment (mumencoder) [inline OpenDreamShared/Compiler/DM/DMParser.cs:162]:
```suggestion

```

Comment (wixoaGit) [reply OpenDreamShared/Compiler/DM/DMParser.cs:162]: If you didn't know, you can make a multiple-line suggestion by click-dragging on the lines when selecting them.

File: OpenDreamShared/Compiler/DM/DMParser.cs (hunk 1 above)
Comment (wixoaGit) [inline OpenDreamShared/Compiler/DM/DMParser.cs]:
```suggestion
                    return statement;
                } finally {
                    _currentPath = oldPath;
                }
```

**Pull request merged**